[김규섭] sprint1

KarubiOhayo opened the pull request:
## 요구사항

### 기본
- [o] 기본 항목 1
- [ ] 기본 항목 2

### 심화
- [o] 심화 항목 1
- [ ] 심화 항목 2

## 주요 변경사항
- 첫 스프린트 미션
- 

## 스크린샷
![image](이미지url)

## 멘토에게
- 수정이나 개선이 필요한 부분이 있다면 알려주세요
-

KarubiOhayo committed:
feat: add User.java

KarubiOhayo committed:
feat: User.java에 멤버 변수 및 생성자, getter 메소드와 update 메소드 추가

KarubiOhayo committed:
feat: Channel.java 추가

KarubiOhayo committed:
feat: Message.java 추가

KarubiOhayo committed:
doc: .gitignore 수정

KarubiOhayo committed:
feat: ChannelService.java, MessageService.java, UserService.java
상기 인터페이스 추가

KarubiOhayo committed:
feat: JCFChannelService.java
JCFMessageService.java
JCFUserService.java
추가 및
UserService.java
ChannelService.java
MessageService.java
수정
각 인터페이스 및 구현체에 CRUD 메서드 구현.

KarubiOhayo committed:
feat: User.java 수정
id, createdAt, updateAt 멤버변수 생성자 내부 초기화 방식 수정
id와 createdAt은 final화
멤버변수 username 추가 및 생성자, getter 메소드, update 메소드 추가.

KarubiOhayo committed:
feat: Channel.java를 수정, User.java와 변경사항은 비슷하나, userCounts와 users(리스트) 변수 추가.
User.java 에는 channels 추가

KarubiOhayo committed:
feat: Message.java 중간 수정

KarubiOhayo committed:
feat: Message.java 중간 수정2

KarubiOhayo committed:
User.java 및 Channel.java 내부 update 메소드 add, delete로 분리
Message.java 수정 및 update 메소드 추가

KarubiOhayo committed:
메인 클래스 제외한 클래스 완성

KarubiOhayo committed:
스프린트 미션 완료

KarubiOhayo committed:
feat: User.java, Channel.java, Message.java 의 toString 수정

KarubiOhayo committed:
feat: UserService.java 인터페이스에 searchById, searchByName 메소드 추가 및 구현체에 해당 메소드 구현.
해당하는 클래스 및 인터페이스에서 일부 메소드명 변경

KarubiOhayo committed:
feat: jcf 패키지 내부 클래스들의 toString 메서드 수정

KarubiOhayo committed:
feat: ChannelService.java 에 searchByName, searchById 메서드 추가.
관련 클래스 및 인터페이스의 메서드명 일부 변경.

KarubiOhayo committed:
feat: JCFUserService.java 의 searchByName 메서드 내부 변수명 수정

KarubiOhayo committed:
feat: MessageService.java 및 하위 구현체에 searchByContent, searchById 메서드 추가.
관련 클래스 및 인터페이스의 메서드명 일부 변경.

KarubiOhayo committed:
feat: MessageService.java 및 하위 구현체에 searchBySenderId 메서드 추가.

the-eris-net commented:
```
.gitignore에 IntelliJ와 관련된 파일이 형상관리 되지 않도록 .idea디렉토리를 추가합니다.
```
해당 첨언이 있었습니다

그리고 현업에서는 .으로 시작하는 파일을 잘 프로젝트에 넣지 않습니다
.으로 시작하는 파일은 리눅스나 맥같은 unix like(유닉스 체계의)운영체제에서 숨김파일이라는 의미입니다.
따라서 보통은 넣지 않는게 일반적입니다.

모든 .으로 시작하는 파일을 넣지 말아야 하는 것은 아닙니다(.gitignore처럼 예외존재)
.idea는 프로젝트에 넣는 파일이 아닙니다
왜냐하면 이 폴더는 intellij의 설정파일이고 조금의 수정으로도 빈번하게 변하여 커밋 내역을 어지럽게 합니다

가급적이면 특별한 이유가 아니라면 빼주시는게 좋습니다.

앞으로는 .으로 시작하는 파일이 일반적으로 넣는 파일인지 검색(GPT 등) 등으로 확인해보고 넣는게 좋습니다

물론 처음 해보셔서 실수 였다고 판단되긴 합니다

the-eris-net commented:
아마 맨처음 git에 추가하면서 그 때는 gitignore가 없었으니까 git add .(전체를 애드하라는 뜻)
을 하셔서 그런거 같습니다

여기서 git rm -r --cached .idea를 하신후 커밋, 푸시하시면 
로컬에는 남고 github에서는 지워질거 같습니다

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/service/ChannelService.java:
```diff
@@ -0,0 +1,22 @@
+import com.sprint.mission.discodeit.entity.Channel;
+
+import java.util.List;
+import java.util.UUID;
```
P3 : 안쓰는 import문은 제거해야합니다
혹시 코드 정리 단축키 쓰시나요?
거기 설정을 수정하면 안쓰는 Import 삭제를 할 수 있습니다

KarubiOhayo replied:
ctrl + alt + o 단축키로 안 쓰는 import문은 제거하고 있긴 한데, 여러가지 수정하는 과정에서 누락 된 부분이 있을 것 같습니다. 근데 지적해주신 ChannelService.java 에서는 List와 UUID 둘 다 사용하고 있는 걸로 알고있는데 다른 부분을 말씀하신 걸까요?

the-eris-net replied:
아 타입 자체가 UUID였군요.. 전 String타입 UUID인줄알고.. 헷갈렸습니다

KarubiOhayo committed:
doc: .idea 실수 수정

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/entity/Channel.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public String toString() {
+        final StringBuilder sb = new StringBuilder("Channel{");
```
혹시 StringBuilder를 사용하신 이유가 있나요?

KarubiOhayo replied:
+로 문자열을 더하면 더할 때 마다 계속 새로운 문자열을 만들기 때문에 메모리 낭비도 심하고 속도도 굉장히 느리다고 알고 있어서 여러가지 면에서 더 효율적인 StringBuilder를 사용하는 것이 좋다고 알고 있습니다.

the-eris-net replied:
P3 : 자바에서는 예상할수 있는 String은 내부적으로 자동으로 StringBuilder처럼 동작합니다
해당 로직이 적용된지는 오래됬습니다.(Java9 이상부터 지원, 2017년 9월 공식 릴리즈)
따라서 String을 성능때문에 StringBuilder를 쓴다는건 이제는 맞지 않는 말입니다.
쉽게 이야기해서 좀 옛날 이야기 입니다

https://openjdk.org/jeps/280
여기에 거기에 대한 이야기를 볼 수 있습니다.

현업에서는 StringBuilder를 반드시 써야하는 상황은
컴파일러가 예측 하지 못하는 상황(for, stream, 메소드로 동적연결)이 아닌
정해진 횟수의 연결은 +를 사용합니다
또한 그게 권장됩니다

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/JavaApplication.java:
```diff
@@ -0,0 +1,67 @@
+
+public class JavaApplication {
+    public static void main(String[] args) {
+        JCFUserService us = new JCFUserService();
```
P4 : 사실 변수의 이름을 극단적으로 줄이는건 좋지 않습니다.
요즘은 자동완성도 좋으니 적절히 userService정도라도 쓰시는게 좋을거 같습니다

KarubiOhayo replied:
아 좋지 않은 습관이었군요...  해당 사항 앞으로도 유의해서 작성하도록 하겠습니다.

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/JavaApplication.java:
```diff
@@ -0,0 +1,67 @@
+public class JavaApplication {
+    public static void main(String[] args) {
+        JCFUserService us = new JCFUserService();
+        JCFChannelService cs = new JCFChannelService();
```
코드작성시 보통은 왼쪽타입은 상위클래스로 적습니다
예를들어 JCFChannelService는 ChannerService cs = new JCFChannelService같은 형식으로요
경우에 따라서 원래의 타입, 혹은 상속중의 타입으로 쓸 수도 있지만 제한적입니다

KarubiOhayo replied:
아 List<T> list = new ArrayList<>() 이런 것 처럼 직접 만든 클래스도 비슷한 방식으로 사용하는게 일반적인가보네요. 생각하지 못한 부분이었습니다. 감사합니다.

the-eris-net replied:
인터페이스의 역할은 외부에 인터페이스에 정의한 메소드를 쓰겠다! 이게 목표입니다
즉 받은 인터페이스 타입만 쓰는게 권장입니다

상황에 따라 차이는 있을 수 있지만 일반적으론 그렇습니다

the-eris-net replied:
물론 main은 일종의 테스트이니 아무렇게 써도 상관은 없지만
실제 코드에서 인터페이스 상속받은 그 타입을 씁니다
하지만 예외 적인 상황들도 있습니다

KarubiOhayo replied:
말씀하신 부분 잘 기억해 두도록 하겠습니다!

KarubiOhayo committed:
feat: 코드리뷰에서 지적받은 부분 수정

toString 오버라이드 메소드의 불필요한 StringBuilder 사용
StringBuilder -> String

극단적으로 짧은 변수명
us -> userService
cs, ms 역시 위와 같이 변경

변수 선언시 왼쪽은 상위 클래스로
JFCUserService userService = new JCFUserService() ->
UserService userService = new JCFUserService()
ChannelService 및 MessageService 역시 위와 같이 변경

KarubiOhayo committed:
feat: 테스트 클래스 JavaApplication에 .java User.java의 멤버변수 Channels와 getChannels 메소드, Channel.java의 멤버변수 Users와 getUsers 메소드 등을 활용해 특정 User가 가입한 Channel들, 특정 Channel에 가입한 User들을 조회하는 테스트 추가.
이 과정에서 User.java와 Channel.java 내부의 get 및 delete 메소드 수정

KarubiOhayo committed:
doc: 테스트 도중 작성한 주석 제거

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/service/jcf/JCFChannelService.java:
```diff
@@ -0,0 +1,73 @@
+                return channel;
+            }
+        }
+        return null;
```
P4 : null을 쓸 때 옵셔널 고려도 좋습니다
다만 불편해지는 경우도 있어서 강하게 요청드리지는 않습니다

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/service/jcf/JCFChannelService.java:
```diff
@@ -0,0 +1,73 @@
+
+public class JCFChannelService implements ChannelService {
+
+    private final List<Channel> data;
```
P1 : 자료구조 List를 쓰면 검색과정에서 엄청난 시간이 걸립니다.
보통 이런 자료구조는 자료주로를 Map을 사용해주서야합니다
이하도 모두 자료구조를 Map으로 변경해주세요

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/service/jcf/JCFMessageService.java:
```diff
@@ -0,0 +1,84 @@
+
+    @Override
+    public List<Message> searchByContent(String content) {
+        List<Message> messages = new ArrayList<>();
```
P1 : List를 사용하다 보니 전체 순회를 하고있습니다. 나중에 map으로 수정하고 나면 해당로직 수정 부탁드립니다

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/JavaApplication.java:
```diff
@@ -0,0 +1,93 @@
+
+        System.out.println(userService.getAllUsers().toString().replace("}, ", "},\n"));
+        System.out.println(channelService.getAllChannels().toString().replace("}, ", "},\n"));
+        System.out.println(messageService.getAllMessages().toString().replace("}, ", "},\n"));
```
테스트 코드에 전반적으로 의도가 들어나지 않습니다.
주석등을 사용해서 의도를 드러내셔야합니다

단순히 이러이런게 나온다 정도는 테스트 무언가를 확인할 수 없습니다.
예를들어 위에서는 리스트를 출력하고 있는데 리스트가 출력이 되야하는건지? 아니면 안출력해야하는건지도 알 수 없습니다.
테스트시 좀 더 의도를 드러내는게 좋습니다

그리고 항상 테스트는 성공과 실패 두가지를 모두 상정해야합니다.
테스트시 어쩔때는 성공하고 어쩔때는 실패한다를 잘 드러나게 해주세요

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/service/jcf/JCFMessageService.java:
```diff
@@ -0,0 +1,84 @@
+
+    @Override
+    public void createMessage(Message message) {
+        data.add(message);
```
P2 : create시 리턴값이 있는게 좋습니다.
만들어졌으면 뭐를 만들어야하는지 보통은 알아야합니다.
특히 db사용시에는 보통 id를 요구하는데 id는 만들어 진 후에 알 수 있기 때문에 반환값에 객체가 보통은 들어갑니다

the-eris-net replied:
P1 : 또한 create시  반드시 user를 체크해야합니다
사실 Message는 Channel도 포함되어야하는데,
둘이 존재해야하만 Message가 존재할 수 있잖아요
따라서 둘의 validation을 체크해줘야합니다

the-eris-net reviewed (approved):
수고하셨습니다. 질문해서 계속 코드를 업그레이드 나가는 자세 좋습니다👍👍👍

the-eris-net commented on src/main/java/com/sprint/mission/discodeit/entity/Channel.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.Objects;
+import java.util.UUID;
+
+public class Channel {
```
P1 : 디스코드 채널에는 설명도 필요하니 description이 추가되어야합니다 또한  공개와 비공개를 결정지을 파라메터가 필요합니다,

각각 description, channelType등으로 만들면 좋을것 같습니다
